sql/api/src/main/scala/org/apache/spark/sql/types/ArrayType.scala

```diff
@@ -43,6 +43,23 @@ object ArrayType extends AbstractDataType {
   override private[spark] def simpleString: String = "array"
 }
 
+/**
+ * Use AbstractArrayType(AbstractDataType) for defining expected types for expression parameters.
+ */
+@Stable
```

Comment:
This is used by expressions to define expected types, so shouldn't be exposed to end-users. It's more like a developer API.

Can we have a new package `org.apache.spark.sql.internal.types` to put this class (with a new file) and also the string collation related classes?